first-party/jib-layer-filter-extension-maven/src/main/java/com/google/cloud/tools/jib/maven/extension/layerfilter/JibLayerFilterExtension.java

```diff
@@ -102,9 +122,120 @@ public ContainerBuildPlan extendContainerBuildPlan(
         .map(FileEntriesLayer.Builder::build)
         .filter(layer -> !layer.getEntries().isEmpty())
         .forEach(newPlanBuilder::addLayer);
+
+    ContainerBuildPlan newPlan = newPlanBuilder.build();
+
+    if (config.get().isCreateParentLayers()) {
+      newPlan = moveParentDepsToNewLayers(newPlan, mavenData, logger);
+    }
+
+    return newPlan;
+  }
+
+  private ContainerBuildPlan moveParentDepsToNewLayers(
+      ContainerBuildPlan buildPlan, MavenData mavenData, ExtensionLogger logger)
+      throws JibPluginExtensionException {
+
+    logger.log(LogLevel.LIFECYCLE, "Moving parent dependencies to new layers.");
+
+    // the key is the expected path for the parent dependency
+    Map<String, Artifact> parentDependencies =
+        getParentDependencies(mavenData)
+            .stream()
+            .map(d -> d.getArtifact())
+            // TODO: configurable?
+            .collect(
+                // TODO: getVersion or getBaseVersion? check in jib, which version it is...
```

Comment:
I tried to look up the difference, and reading this SO, I think `getVersion()` is what we want.

Reply:
Thank you for clatrification. Will leave it open until I have looked into it a little bit more.

Reply:
@chanseokoh , I am afraid this is not correct.  I added the follwoing dependency to my **parent** pom:
```
<dependency>
			<groupId>commons-io</groupId>
			<version>2.9.0-SNAPSHOT</version>
			<artifactId>commons-io</artifactId>
		</dependency>
```

But it was not moved to a new "snapshots dependencies-parent" layer. Instead I got in log:

`Keep /app/libs/commons-io-2.9.0-SNAPSHOT.jar in snapshot dependencies.`

So it seems, that jib has used the base version when creating the layers. 

And my  getVersion() is the "unique snapshot  version" with timestamp. So they do not match:

`Dependency from parent not found: /app/libs/commons-io-2.9.0-20210217.161556-2.jar`

Reply:
I did some more debugging in jib...  So, this is the place, where the commons-io jar is added:

https://github.com/GoogleContainerTools/jib/blob/v2.8.0-maven/jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/MavenProjectProperties.java#L340

```
classifiedDependencies
            .get(LayerType.SNAPSHOT_DEPENDENCIES)
            .add(artifact.getFile().toPath());
```
Here artifact.getFile() gives us the "-SNAPSHOT.jar".

I cannot prove this by Maven's javadoc or so, but this is also the filename I see usually in my IDE (Eclipse), when unfolding "Maven Dependencies". There won't be ever filenames with timestamps there...

Reply:
> So it seems, that jib has used the base version when creating the layers.

I don't think this is true.

Here's my understanding about SNAPSHOT artifacts:

### Snapshot artifacts published to a local Maven repository
   - They don't have timestamps, and the file ends with `-SNAPSHOT`, without a timestamp.
   - Therefore, `getBaseVersion()` and `getVersion()` returns the same string `-SNAPSHOT`, without timestamps.
   - They are what this SO answer calls "non-unique." `mvn install` simply overwrites `-SNAPSHOT` file.
      > [Non-unique snapshot] ends with `-SNAPSHOT` and is generated when you work with projects on your own system. A `mvn install` will copy the artifact with the -SNAPSHOT version to the local repository.

### Snapshot artifacts published to a remote repository
   - Whenever a snapshot artifact is uploaded, they get timestamped. This is because Maven repositories are immutable by its design. Once an artifact when live, it stays there forever to ensure reproducible builds. It's no exception for published snapshot artifacts. For example, you can see here [multiple SNAPSHOT artifacts with the same version are published with different timestamps](https://maven.wso2.org/nexus/content/repositories/Apache/commons-io/commons-io/2.6-SNAPSHOT/).
   - Against a resolved artifact, `getBaseVersion()` returns `-SNAPSHOT`, while `getVersion()` returns `-SNAPSHOT` with a timestamp.
   - Each snapshot artifact is unique, and they are what the SO answer calls "unique."
      > If you use SNAPSHOTs from a remote repository, then those artifacts are timestamped (ends with a human friendly representation of a timestamp), which makes them unique. This has to do with the way they are uploaded to the remote repository.

And I believe `artifact.getFile()` always return the actual file. Note it may or may not have a timestamp; it depends on whether the artifact is resolved against a local repo or a remote repo,

So, here's what I suspect has happened in your experiment:
- When Jib ran, `commons-io` was taken from a local repository. The file doesn't have a timestamp.
- When the extension ran, `commons-io` was resolved against a remote snapshot repository.

In any case, I think both `getBaseVersion()` and `getVersion()` can break. I think it happens to be that in your case, `getBaseVersion()` worked. However, it could be the other way around.

Some options I can think of:
- Approximate the matching so that the extension always moves snapshot versions, whether they are timestamped or not.
- Figure out the exact artifact files that Jib used. Looking at `MavenProjectProperties`, I think the artifact list can be retrieved by `MavenProject.getArtifacts()`.

Reply:
@chanseokoh , thank you for explanation. I will see how I can further improve the "snapshot matching". In any case I will have the issue of having to match a filename or path with a Maven `Artifact`.  The cleanest way to do this would to have the originating `Artifact` as an optional field in `FileEntry`. But this is of course another jib change. An we would probalby need a to have a subclass like `MavenFileEntry`, since `FileEntry` must be independent from Maven.
For the time being, I will look for a better way of matching without that kind of jib change.

Reply:
@chanseokoh , I think I have it with https://github.com/GoogleContainerTools/jib-extensions/pull/76/commits/d974582e3e6fed55c0235e7cafca604945184765.
I now match the source file path of the resolved `Artifact` with the source file path of the `FileEntry`.
I don't make any assumptions about the structure of those path.
The only assumption that is left is that Maven will resolve to the samefile path, if the same artifact coordinates are resolved multiple times within the same build / session.
Furthermore this also resolves the potential issue with duplicate artifact ids that I had before ("duplicates").